### PR TITLE
Add JSON parsing for Hero and Villain CCs

### PR DESCRIPTION
**Pull request by sjzhu**
## What?
Adding the ability to do JSON parse/output on Hero and Villain CCs!

It makes the JSON parse code a little more complicated, but I'm not really sure how to do this sort of thing in a "clean" way. It's gonna need to be made even more complicated once we support events and principles and the like.

## How was it tested?
Locally: 
image
image

## Reviewers
@Colcoction @alexdarling @Fosuke

### CODE REVIEW

File: hero-character-card-front/index.html [hunk 1]
```diff
@@ -154,7 +154,7 @@ <h1>Hero Character Card Front</h1>
       </div>
 
       <div class="inputGroup">
-        <label for="inputTitle">Power Name</label>
```
Comment (alexdarling) [inline hero-character-card-front/index.html]: oop undercooked pasta strikes again

File: hero-character-card-front/script.js [hunk 2]
```diff
@@ -67,7 +67,8 @@ $('#inputImageFile').on('input', function (e) {
 let loadedUserImages = {
```
Comment (alexdarling) [inline hero-character-card-front/script.js]: Not sure how much scope this adds, but could we define this as a map, or at least write variables for the keys like `const BACKGROUND_ART = "backgroundArt";`? I'm a little concerned at how many `loadedUserImages["stringThatMightHaveATypoWeCouldMiss"]` statements are showing up above.

File: common-after.js [hunk 3]
```diff
@@ -608,32 +610,192 @@ function parseJSONData(data) {
...
+    $('#inputPowerName').val(data.PowerName);
+  } else {
+    $('#inputPowerName').val('');
+  }
+  if('NemesisIconURL' in data && data.NemesisIconURL.length != 0) {
+    loadedUserImages['nemesisIcon'] = new Image();
+    loadedUserImages['nemesisIcon'].src = data.NemesisIconURL;
+    loadedUserImages['nemesisIcon'].onload = function () {
+      // Once the Image has loaded, redraw the canvas so it immediately appears
+      drawCardCanvas();
```
Comment (alexdarling) [inline common-after.js:725]: We're already calling `drawCardCanvas()` at the bottom of the function, and (I think) JS is just going to execute all of the logic here at once before it update the DOM, so I don't think we need any additional calls to `drawCardCanvas()` in places like here / the background art / etc.

**Review by alexdarling — CHANGES_REQUESTED** [review]
2 requests of you & 1 joke comment

File: common-after.js (hunk 3 above)
Comment (sjzhu) [reply common-after.js:725]: Hmmm, so if I remove all of these drawCardCanvas calls (except for the last one on line 735), once the JSON is parsed, the images are all still blank. I think the drawCardCanvas call is executed before the images are loaded, and so they don't get rendered properly. If I make any changes after the JSON is parsed, then the images render. 
Is there a way we can make a listener that waits for all the images to be loaded and then calls drawCardCanvas once that happens?

Comment (alexdarling) [reply common-after.js:725]: Something about that seems off and I don't know what it is. We're running everything in order in our single thread, so calling the draw here vs above shouldn't be an issue... at least I think. Could you post the JSON you're using for your test so I can repro on my end?

One thing we could try is a deferred function to force the `drawCardCanvas()` call into a new thread:

```javascript
setTimeout(drawCardCanvas, 0);
```

My gut says that won't work, but it's probably worth a try.

Comment (sjzhu) [reply common-after.js:725]: Here's an  example of a JSON that doesn't load immediately: https://gist.github.com/sjzhu/85698581df60ed4bb3fcc3aeae6d36ed
image
After changing the HP value in the form field:
image

File: hero-character-card-front/script.js (hunk 2 above)
Comment (sjzhu) [reply hero-character-card-front/script.js]: Very reasonable concern! I'm working on this in a separate branch now:  https://github.com/sjzhu/unitys-workshop/compare/sjzhu--add-cc-json...sjzhu:unitys-workshop:sjzhu--make-globals-for-image-purposes?expand=1

Comment (sjzhu) [reply hero-character-card-front/script.js]: Almost a year later, I return to this PR and have made a lot of global constants 😂

**Comment by sjzhu** [comment]
Add Villain CCs to this PR because it made sense! Test image: 
vcc-json-test

**Comment by sjzhu** [comment]
Screenshot with more context: 
image

**Comment by sjzhu** [comment]
The comment screenshots are no longer accurate, the main PR message has the final layout!

**Pull request merged**